Feature/lc gitlab ci

adrienbernede opened the pull request:
**This PR is a feature. It does the following:**
  - **Reformat host-config files (split shareable and project specific), rename host-config-files (for use as regex)**
  - **Introduces CI pipeline description for GitLab in LC context** 

This feature aims at introducing a first GitLab CI pipeline designed for the LC GitLab instance.
The additional constraint was to design this pipeline so that we could share as much configuration as possible with other projects.

## First Action

The idea is to separate project specific and shareable configuration.

As for now, shareable information mainly consist in toolchain configuration, such as compiler paths, and is placed in the `.gitlab/conf` directory. More generally, the aim is to have .gitlab containing non-project specific gitlab configuration so that it can be easily moved elsewhere later on (submodule ?).

Project specific configuration, such as compiling flags and target release identification, remains at the same place.

Shareable and project configuration are linked by the directory layout which has been unified : `system_type/toolchain`

## Second Action

The `.gitlab-ci.yml` file is used by GitLab CI as the main configuration file. Using inclusion of files (`include:`) and job properties extensions (`extends:`) the pipeline configuration is split in several places in a modular manner.

`.gitlab-ci.yml` containins project specific configuration of the pipeline.

`.gitlab/ci` contains shareable configuration, such as configuration of jobs for different clusters.

In addition, an allocation mechanism is introduced so that on quartz only one resource allocation is required. Note that this allocation is currently using the pdebug queue.

## Known Limits

Currently, this feature is not perfect, and requires some discussion and choices to be made.
In fact, its main drawback is that the builds generated in the pipeline do not always correspond to the configuration described in build scripts.[EDIT: no longer true with latest changes]

### Main change
Host-config files have been split in two parts. The shareable/toolchain configuration, and the project specific configuration which adapts the config to the toolchain.

### The choice
I see two main ways to write the project specific host-configs: flat (umpire) and modular (raja). 
 - In the flat mode, there is one host-config file per configuration, and it is straightforward to design a script which can build any of them. 
 - In the modular mode, configuration is shared among similar targets, writing a script for any configuration is less obvious.
In this PR, I wrote a script to launch any build, using project host-config names as a regex to identify the one matching the selected toolchain.

However, there is in RAJA a third level of configuration: at script time. This is not handled yet. And I would like it to be discussed. [EDIT: no longer true with latest changes -> introduce extra config parameter, allowing to create "tuned" versions of configurations, like when using "omptarget"]

However, It may not delay the treatment of this PR which is functional and IMO adds value to the testing framework as-is.

davidbeckingsale commented:
This is a great start, thanks @adrienbernede !

adrienbernede commented:
Thanks ! 
Feel free to ask questions.

adrienbernede commented:
By the way, if I could be added as a developer (I don’t know the exact requirement), I would then be able to configure the RAJA repo on LC GitLab CI.

rchen20 commented on scripts/gitlab/bulld_and_test.sh:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
```
I'm guessing this file was created for your own testing purposes and won't be included in the final PR?

davidbeckingsale replied:
I think this script is used by the GitLab jobs...

rchen20 replied:
There is another file scripts/gitlab/build_and_test.sh that is very similar, which I assumed was the actual GitLab script.

davidbeckingsale replied:
Whoops, I misread this filename.

adrienbernede replied:
Good catch !
It copy created by mistake (typo). I thought I deleted it, but I may have resurrected it while cleaning the branch.
Corrected.
Thanks

rhornung67 commented:
@adrienbernede I added you as a collaborator with write access to the repo. Please let me or @davidbeckingsale know if you encounter access issues.

adrienbernede commented:
@rhornung67 Thanks for adding me. I’ll configure the LC GitLab CI repo.

I forced push the fix removing "scripts/gitlab/build_and_test.sh". It dismissed all the reviews...
If there is another way, please let me know so that it doesn’t happen next time.

rchen20 commented on .gitlab/conf/host-configs/blueos_3_ppc64le_ib/nvcc_9_2_xl_2019_04_19.cmake:
```diff
@@ -0,0 +1,11 @@
+
+set(CMAKE_CXX_COMPILER "/usr/tce/packages/xl/xl-2019.04.19/bin/xlc++_r" CACHE PATH "" )
+
+set(CUDA_TOOLKIT_ROOT_DIR "/usr/tce/packages/cuda/cuda-9.2.148" CACHE PATH "" )
```
@adrienbernede Everything looks great! This is a good start.

"However, there is in RAJA a third level of configuration: at script time. This is not handled yet. And I would like it to be discussed."

I'm not entirely sure what you mean by this, but in general I think the RAJA level of modularity is good. In fact, we could attempt to make it even _more_ modular by separating the combinations of compiler paths. For example, this particular file could be chopped into an "nvcc_9_2.cmake" and an "xl_2019_04_19.cmake". This would become rather complex though, because we would also need to separate host-configs/.../nvcc_X_xl_X.cmake into 2 files as well, and the gitlab/build_and_test.sh regex would become more difficult.

adrienbernede replied:
"However, there is in RAJA a third level of configuration: at script time. This is not handled yet. And I would like it to be discussed."

What I meant there is that in RAJA build scripts the configuration is determined using 3 sources of information:
- the toolchain config files (in .gitlab/conf/host-configs),
- the project config files (in host-configs),
- and the script itself, providing definitions in the CMake command line. 
This last "level of configuration" is not present in the generic `scripts/gitlab/build_and_test.sh`. This means that the CI builds are not the same as the scripted builds at the moment. This is important to notice, and I would like to discuss the ways to address this.

You are right, the current design is a trade-off between modularity and simplicity. What I envision next is to generate or retrieve the toolchain config-files from a list of desired toolchain. This way we could have a single, let’s say yaml, file were the configurations are listed in a compact and readable way.

rchen20 replied:
If I understand you correctly, you're saying that cmake options like "-DENABLE_OPENMP=On" or "-DENABLE_CUDA=Off" (and all such combinations) are difficult to encode in gitlab/build_and_test.sh? This would indeed be a problem.

One way to solve this (possibly difficult) would be to use regex rules. For instance, if we detect nvcc in the config then include "-DENABLE_CUDA=On", and depending on the nvcc version number set "-DCUDA_ARCH=sm_something". This will devolve into a build_and_test.sh that is difficult to grok though.

We could also have separate build_and_test.sh's for different general cases, for example:

- build_and_test_nvcc10.sh
- build_and_test_nvcc9.sh
- build_and_test_nvcc8.sh
- build_and_test.sh
- build_and_test_omp.sh
- build_and_test_omptarget.sh

Each of these would have static cmake lines, and read in their corresponding test cases.

adrienbernede replied:
I think the examples you gave are not the ones that would cause the most troubles.
In fact, _if there is a simple rule_ that dictates with which toolchain we want OpenMP and or Cuda, and with which nvcc version to associate a sm_xx arch, then we could deal with it.

For example, the build_and_test.sh script currently expects a unique matching host-configs file. But we could imagine using all the matching ones.  

For example, both `nvcc_X_clang_X.cmake` and `nvcc_9_2_X.cmake` would match nvcc_9_2_clang_6_0_0.cmake. And `nvcc_9_2_X.cmake` would set `-DCUDA_ARCH=sm_something`.

However, this proposal cannot deal with two different configurations applying to the same target (Debug/Release/EA). I have some twisted idea about how to do that, though.

adrienbernede commented:
LGTM

rhornung67 commented:
Let's hold off on merging this until after the holidays so everyone who wants to can weigh in on it. @davidbeckingsale @artv3 @trws @ajkunen @rchen20 @mdavis36 @MrBurmark please look this over and let us know if you are OK with it and/or have any questions or concerns.  Thanks.

adrienbernede commented:
@rhornung67 Sure thing!
The feature would also use an update, I was just testing mirroring with Gitlab.

adrienbernede commented:
Updated the PR with latest changes in configurations.

Added support for a suffix in host-configs file naming. This allows to tune the configuration for specific use like omptargets. (IMHO, scripts/gitlab/build_and_test.sh could replace the usage of multiple scripts at this stage, which would ease maintenance).

i.e. on LC:
`> CONFIGURATION=clang_coral_2018_08_08__omptarget scripts/gitlab/build_and_test.sh`
Will do the same as, 
`> scripts/lc-builds/blueos_clang-coral-2018.08.08_omptarget.sh`
because all RAJA specific parameters, including those due to the use of openmp targets, are now included in
`host-configs/blueos_3_ppc64le_ib/clang_coral_2018_08_08__omptarget.cmake`
and the corresponding toolchain is defined in
`.gitlab/conf/host-configs/blueos_3_ppc64le_ib/clang_coral_2018_08_08`
(which could be exported to radiuss-ci, cf https://github.com/LLNL/Umpire/pull/294)

Thus, testing a new parameter for this configuration simply requires to duplicate the host-config file, and change the suffix:
`host-configs/blueos_3_ppc64le_ib/clang_coral_2018_08_08__mytest.cmake`
is then callable using:
`> CONFIGURATION=clang_coral_2018_08_08__mytest scripts/gitlab/build_and_test.sh`

Again, this is a proposal, comments are welcome. If it gets accepted, I will add tests and the possibility to print the full configuration (dry run).

adrienbernede commented:
LGTM

adrienbernede commented:
@rhornung67 @davidbeckingsale 

The longer this branch lives, the more conflicts it gets because it impacts host-configs and build scripts. Updating is painful, but I did it 10 days ago with the hope to trigger a discussion. Now its out-dated again.

I would like the changes to be discussed so that we can agree on what to do and hopefully avoid another painful updates because this branch lived too long.

Thanks

adrienbernede commented:
PR move to #737 to get feedback from Gitlab CI.
